### PR TITLE
Reinstate the old zpool read label logic as a fallback.

### PR DESCRIPTION
**Pull request by rincebrain**
### Motivation and Context
#12032 

### Description
I landed a verbatim copy of the original zpool_read_label as "zpool_read_label_slow", then reshuffled the logic in
zpool_read_label to call zpool_read_label_slow if any of the AIO operations returned EOPNOTSUPP, which is what 
FBSD spits back when you try AIO on FDs where it thinks it's "unsafe".

(The reshuffling was necessary because errno after this happens isn't (always?) EOPNOTSUPP, at least some of the time it's EIO; I landed the whole old function rather than integrating the new and old logic into one function after seeing that the diff between them was almost all of the function body, so it would have just been two entirely separate copies of the logic inside anyway.)

### How Has This Been Tested?
* Tried importing and creating a pool over NFS on FBSD 13/x86_64 before and after this change with `vfs.aio.enable_unsafe=0` (it fails before and works after).
* Tried pool creation and importing on Debian bullseye/x86_64
* Tried pool importing on CentOS 7/x86_64
* Tried pool importing on Ubuntu 18.04/x86_64

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by behlendorf** [comment]
cc: @asomers

File: lib/libzutil/zutil_import.c [hunk 1]
```diff
@@ -929,22 +1006,35 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
-				int r = aio_error(&aiocbs[l]);
-				if (r != EINVAL)
+		int tmpret = -1;
+
+		for (l = 0; l < VDEV_LABELS; l++) {
+			errno = 0;
+			int r = aio_error(&aiocbs[l]);
+			if (r != 0) {
+				if (r == EOPNOTSUPP) {
+					/*
```
Comment (asomers) [inline lib/libzutil/zutil_import.c:1016]: You still need to call `aio_return` here for all operations.  Otherwise you'll leak AIO resources.

**Review by asomers — CHANGES_REQUESTED** [review]
I think this is a good idea.  It never occurred to me that somebody might try to create a pool over NFS.

File: lib/libzutil/zutil_import.c [hunk 2]
```diff
@@ -929,22 +1006,35 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
-		if (errno == EAGAIN || errno == EINTR || errno == EIO) {
-			/*
-			 * A portion of the requests may have been submitted.
-			 * Clean them up.
-			 */
-			for (l = 0; l < VDEV_LABELS; l++) {
-				errno = 0;
-				int r = aio_error(&aiocbs[l]);
-				if (r != EINVAL)
+		int tmpret = -1;
```
Comment (freqlabs) [inline lib/libzutil/zutil_import.c:1009]: We could reuse `error` instead of introducing a new return variable.

File: lib/libzutil/zutil_import.c (hunk 1 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c:1016]: Sure.

If I'm understanding this correctly, doesn't the need to always call aio_return imply that the old code also leaked resources if you got back something that wasn't (EINTR|EAGAIN|EIO)? (I'm finding it hard to find documentation on when it's appropriate to call versus not - the closest thing I found was a macOS man page which said you should always do it...

File: lib/libzutil/zutil_import.c [hunk 3]
```diff
@@ -929,22 +1006,35 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
+
+		for (l = 0; l < VDEV_LABELS; l++) {
+			errno = 0;
+			int r = aio_error(&aiocbs[l]);
```
[the remark this answers is not in the document]
Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: Good idea. Done!

File: lib/libzutil/zutil_import.c (hunk 2 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: Good point. Done!

File: lib/libzutil/zutil_import.c (hunk 1 above)
Comment (asomers) [reply lib/libzutil/zutil_import.c]: No.  You only need to call `aio_return` if the aiocb was actually started.

File: lib/libzutil/zutil_import.c [hunk 4]
```diff
@@ -1006,35 +1006,39 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
 
 	if (lio_listio(LIO_WAIT, aiocbps, VDEV_LABELS, NULL) != 0) {
 		int saved_errno = errno;
-		int tmpret = -1;
+		errno = 0;
+		int do_slow = 0;
```
Comment (nabijaczleweli) [inline lib/libzutil/zutil_import.c:1010]: this would probably prefer to be a `boolean_t`

File: lib/libzutil/zutil_import.c [hunk 5]
```diff
@@ -929,22 +1006,39 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
+					do_slow = 1;
+				case EIO:
+				case EINTR:
+				case EAGAIN:
+					/*
+					 * A portion of the requests may have
+					 * been submitted. Clean them up.
+					 */
 					(void) aio_return(&aiocbs[l]);
 			}
```
Comment (asomers) [inline lib/libzutil/zutil_import.c:1029]: You need a default statement here.  You should probably `assert` if `aio_error` returns `EINPROGRESS`.  That shouldn't be possible with `LIO_WAIT`.  Or perhaps just print a warning.  In either case, falling back to the slow path might be appropriate for unknown errors.

File: lib/libzutil/zutil_import.c [hunk 6]
```diff
@@ -929,22 +1006,39 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
-				int r = aio_error(&aiocbs[l]);
-				if (r != EINVAL)
+		errno = 0;
+		int do_slow = 0;
+		error = -1;
+
+		for (l = 0; l < VDEV_LABELS; l++) {
+			switch (aio_error(&aiocbs[l])) {
+				case 0:
+					break;
```
Comment (asomers) [inline lib/libzutil/zutil_import.c:1016]: You need to `aio_return` in the successful case.

File: lib/libzutil/zutil_import.c [hunk 7]
```diff
@@ -929,22 +1006,39 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
+				case -1:
+					break;
+				case EOPNOTSUPP:
+					do_slow = 1;
+				case EIO:
+				case EINTR:
+				case EAGAIN:
+					/*
+					 * A portion of the requests may have
```
Comment (asomers) [inline lib/libzutil/zutil_import.c:1025]: I would move this comment up.  It applies whenever `lio_listio` returns nonzero.

File: lib/libzutil/zutil_import.c [hunk 8]
```diff
@@ -929,22 +1006,39 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
+
+		for (l = 0; l < VDEV_LABELS; l++) {
+			switch (aio_error(&aiocbs[l])) {
+				case 0:
+					break;
+				case -1:
+					break;
+				case EOPNOTSUPP:
+					do_slow = 1;
+				case EIO:
```
Comment (asomers) [inline lib/libzutil/zutil_import.c:1021]: The EIO, EINTR, and EAGAIN error codes might be returned by `lio_listio`, but they should not be returned by `aio_error`.

File: lib/libzutil/zutil_import.c [hunk 9]
```diff
@@ -929,22 +1006,39 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
+		errno = 0;
+		int do_slow = 0;
+		error = -1;
+
+		for (l = 0; l < VDEV_LABELS; l++) {
+			switch (aio_error(&aiocbs[l])) {
+				case 0:
+					break;
+				case -1:
+					break;
```
Comment (asomers) [inline lib/libzutil/zutil_import.c:1018]: Programming defensively, I would assert that the error is EINVAL here.

File: lib/libzutil/zutil_import.c [hunk 10]
```diff
@@ -929,22 +1006,39 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
+			 * At least some IO involved access unsafe-for-AIO
+			 * files. Let's try again, without AIO this time.
+			 */
+			error = zpool_read_label_slow(fd, config, num_labels);
+			saved_errno = errno;
+		}
 		free(labels);
 		errno = saved_errno;
-		return (-1);
+		return (error);
```
Comment (asomers) [inline lib/libzutil/zutil_import.c:1041]: Why are you changing the return value?  That doesn't seem related to the rest of the PR.

**Review by asomers — CHANGES_REQUESTED** [review]

File: lib/libzutil/zutil_import.c (hunk 10 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c:1041]: The goal was to not return -1 if zpool_read_label_slow gets called and returns 0, since we did succeed at the purpose of the function.

Comment (asomers) [reply lib/libzutil/zutil_import.c:1041]: Ok, and `zpool_read_label_slow` will only return either 0, or -1.  I get it.  I thought that you had changed it to return an errno or something, but now I see that it is not the case.

File: lib/libzutil/zutil_import.c (hunk 8 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c:1021]: The documentation for aio_error on Linux says it can spit out ~anything that read could, if the operation failed, unless I'm misreading it?

Comment (asomers) [reply lib/libzutil/zutil_import.c:1021]: An EIO from aio_read means something very different from an EIO from lio_listio.  Only in the latter case does it mean that "A portion of the requests may have been submitted".  If aio_read returns EIO, it means that the disk actually wasn't readable.  Also, you should probably handle `ENOSYS` just like `EOPNOTSUPP`.  That won't happen on FreeBSD, but might on other platforms.

Comment (rincebrain) [reply lib/libzutil/zutil_import.c:1030]: Sure; I think I've got a reasonable workflow for this done; let me give it a whirl through some testing before I push it...

File: lib/libzutil/zutil_import.c [hunk 11]
```diff
@@ -1005,26 +1005,32 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
 	}
 
 	if (lio_listio(LIO_WAIT, aiocbps, VDEV_LABELS, NULL) != 0) {
+		/*
+		 * A portion of the requests may have
+		 * been submitted. Clean them up.
+		 */
 		int saved_errno = errno;
+		boolean_t do_slow = 0;
```
Comment (nabijaczleweli) [inline lib/libzutil/zutil_import.c:1013]: if it's a boolean_t then you probably want to assign B_FALSE and B_TRUE as well, rather than 0/1

Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: Very good point, done.

File: lib/libzutil/zutil_import.c (hunk 9 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: Good idea, done.

File: lib/libzutil/zutil_import.c (hunk 7 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: Makes sense. Done.

File: lib/libzutil/zutil_import.c (hunk 6 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: I think I've done aio_return in all the right cases in this latest refactoring? PTAL and let me know if I've missed any cases; I think this represents my best understanding of when it's necessary.

File: lib/libzutil/zutil_import.c (hunk 5 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c:1034]: Sure, I opted to ASSERT on it because it looks like printing an error in the way done elsewhere in zutil would require changing the function signature, which seemed like a nonstarter.

File: lib/libzutil/zutil_import.c (hunk 4 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: As you've already noticed, done.

File: lib/libzutil/zutil_import.c [hunk 12]
```diff
@@ -1005,27 +1005,32 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
 	}
 
 	if (lio_listio(LIO_WAIT, aiocbps, VDEV_LABELS, NULL) != 0) {
+		/*
+		 * A portion of the requests may have
```
Comment (asomers) [inline lib/libzutil/zutil_import.c:1009]: Now the comment is too high.  It should go just beneath the check for EAGAIN, EINTR, and EIO.

File: lib/libzutil/zutil_import.c (hunk 8 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: This reply is kind of redundant with my one above, but I have refactored how this code flow works in response to this; if you have a chance, PTAL and let me know if I've missed something again.

File: lib/libzutil/zutil_import.c [hunk 13]
```diff
@@ -1005,27 +1005,32 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
-					do_slow = 1;
-				case EIO:
-				case EINTR:
-				case EAGAIN:
-					/*
-					 * A portion of the requests may have
-					 * been submitted. Clean them up.
-					 */
-					(void) aio_return(&aiocbs[l]);
+		if (saved_errno == EAGAIN || saved_errno == EINTR ||
```
Comment (asomers) [inline lib/libzutil/zutil_import.c:1016]: At this point saved_errno and errno should be identical.  So you can eliminate the line wrap by using plain `errno`.

File: lib/libzutil/zutil_import.c [hunk 14]
```diff
@@ -929,22 +1006,43 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
+			 * A portion of the requests may have
+			 * been submitted. Clean them up.
 			 */
 			for (l = 0; l < VDEV_LABELS; l++) {
 				errno = 0;
-				int r = aio_error(&aiocbs[l]);
-				if (r != EINVAL)
-					(void) aio_return(&aiocbs[l]);
+				switch (aio_error(&aiocbs[l])) {
+					case EINVAL:
```
Comment (freqlabs) [inline lib/libzutil/zutil_import.c:1020]: The case labels can be at the same indent level as the switch.

File: lib/libzutil/zutil_import.c (hunk 12 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: Ah, I see, I'm sure I had good reason to put it there before, but looking at your prior statement now, I'm unsure how I could have misinterpreted it.

Fixed.

File: lib/libzutil/zutil_import.c (hunk 13 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: Good point. Corrected.

File: lib/libzutil/zutil_import.c [hunk 15]
```diff
@@ -929,22 +1006,43 @@ zpool_read_label(int fd, nvlist_t **config, int *num_labels)
...
 		int saved_errno = errno;
+		boolean_t do_slow = B_FALSE;
+		error = -1;
 
 		if (errno == EAGAIN || errno == EINTR || errno == EIO) {
 			/*
-			 * A portion of the requests may have been submitted.
-			 * Clean them up.
+			 * A portion of the requests may have
+			 * been submitted. Clean them up.
```
Comment (asomers) [inline lib/libzutil/zutil_import.c:1015]: Having moved this comment back to where it started, it now looks like you unnecessarily wrapped it.

File: lib/libzutil/zutil_import.c (hunk 14 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: Works for me. Done.

File: lib/libzutil/zutil_import.c (hunk 15 above)
Comment (rincebrain) [reply lib/libzutil/zutil_import.c]: Sounds like something I'd do, moving things around the long way only to wind up in the identical place.

Done.

**Review by freqlabs — APPROVED** [review]

**Comment by freqlabs** [comment]
Now we just need to squash all those commits and make sure the final commit is signed off on to satisfy checkstyle.

**Comment by rincebrain** [comment]
Hm, the merge with master plus the rebase has it reporting that I committed all the commits that were merged from master. LMK if I should have done that some other way, and I'll go rewrite history appropriately.

**Comment by freqlabs** [comment]
No worries, sorted it out for you.

**Comment by rincebrain** [comment]
For my future reference, how did you clean it up?

(...also, I didn't realize contributors could force push to others' branches.)

**Review by asomers — APPROVED** [review]

**Comment by freqlabs** [comment]
> For my future reference, how did you clean it up?

`git rebase -i master` making sure only the one commit was being picked.

> (...also, I didn't realize contributors could force push to others' branches.)

There's a checkbox when you open a PR that allows contributors to make changes.
